src/background/StreamSessionManager.ts

```diff
@@ -0,0 +1,373 @@
+                                user.connection?.send(syncStateMessage);
+                            }
+                        });
+                    }, 500);
+                }
+            }
+        });
+    }
+
+    /**
+     * Determines the new leader id after the leader disconnects. Assumes the previous leader has been removed from sessionState
+     */
+    private determineNewLeaderId(): string {
+        for (const userId of this.sessionState.sessionJoinOrder) {
+            const user = this.sessionState.sessionUsers.get(userId);
+            if (user?.permissions.canControlVideo) {
+                return user.id;
+            }
+        }
+
+        // If we haven't found any other users who can control video, default to the oldest in the join order
+        return this.sessionState.sessionJoinOrder[0];
+    }
+
+    /**
+     * Handler for state sync request messages
+     */
+    private handleStateSyncRequestMessage(): void {
+        console.warn("state sync requested");
+
+        if (this.sessionState.leaderId !== this.currentUserId) {
+            console.warn(`Received sync state request on ${this.currentUserId} but leader is: ${this.sessionState.leaderId}`);
+            return;
+        }
+
+        const stateSyncMessage = this.generateSyncStateMessageFromCurrentState();
+        this.sessionState.sessionUsers.forEach(user => {
+            if (user.id !== this.currentUserId) {
+                console.warn(`Sending sync state to: ${user.id}`);
+                user.connection?.send(stateSyncMessage);
+            }
+        });
+    }
+
+    /**
+     * Handler for state sync messages
+     * @param message The state sync message to process
+     */
+    private handleStateSyncMessage(message: IStateSyncMessage): void {
+        console.warn("handling state sync");
+        const numUsersBefore = this.sessionState.sessionUsers.size;
+        // List of ids used for cleaning up users list after updates
+        const newListIds: string[] = [];
+
+        for (const user of message.state.sessionUsers) {
+            const existingUser = this.sessionState.sessionUsers.get(user.id);
+            const existingUserPeerJsConnection = existingUser?.connection;
+
+            const newUserObj: IUserConnection = {
+                id: user.id,
+                displayName: user.displayName,
+                connection: existingUserPeerJsConnection,
+                permissions: user.permissions
+            };
+            this.sessionState.sessionUsers.set(user.id, newUserObj);
+            newListIds.push(user.id);
+        }
+
+        // We've set the new users, but we need to clean up any users that weren't in the message
+        this.sessionState.sessionUsers.forEach(user => {
+            if (newListIds.indexOf(user.id) < 0) {
+                // User entry wasn't in the list of users we got from the host
+                this.sessionState.sessionUsers.delete(user.id);
```

Comment:
Would we also want to disconnect from them?

Reply:
Ah good catch. Should be a quick fix

Reply:
Actually I think connects and disconnects should be handled on a per user basis instead of a result of state syncs. It works because everyone is connected to everyone. That way we can avoid some race conditions where some people have disconnected from the user who left but some are still waiting for or processing the message